ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/FSharpTypingAssist.fs

```diff
@@ -736,49 +739,57 @@ type FSharpTypingAssist(lifetime, dependencies) as this =
         let mutable allowedTokens = null
         if not (tryDeindentTokens.TryGetValue(lexer.TokenType, &allowedTokens)) then false else
 
-        let prevTokenEnd =
+        let tokenBeforeKeywordEnd =
             use cookie = LexerStateCookie.Create(lexer)
             lexer.Advance(-1)
             while isIgnored lexer.TokenType do
                 lexer.Advance(-1)
             lexer.TokenEnd
 
-        match x.GetFSharpTree(textControl) with
-        | None -> false
-        | Some parseTree ->
-
-        let document = textControl.Document
-        let visitor =
-            { new SyntaxVisitorBase<_>() with
-                member x.VisitExpr(_, _, defaultTraverse, expr) =
-                    match expr with
-
-                    // if expr then {caret}
-                    // No info is available after error recovery, we check that range is surrounded with if ... then.
-                    | SynExpr.FromParseError (ExprRange range, _)
-
-                    // while expr do {caret}
-                    | SynExpr.While (_, ExprRange range, _, _) when
-                            getEndOffset document range = prevTokenEnd ->
-                        Some range
-
-                    // for i = ... to ... do {caret}
-                    | SynExpr.For (_, _, IdentRange range, _, _, _, _, ExprRange lastRange, _)
+        let expr: IFSharpExpression = getNode textControl lexer.TokenStart
+        if isNull expr then false else
 
-                    // for ... in ... do {caret}
-                    | SynExpr.ForEach (_, _, _, _, PatRange range, ExprRange lastRange, _, _) when
-                            getEndOffset document lastRange = prevTokenEnd ->
-                        Some range
+        let indentOffset =
+            let inline tryGetIndentOffset (exprBeforeKeyword: IFSharpExpression) (indentAnchor: ITreeNode) =
+                if isNotNull exprBeforeKeyword && isNotNull indentAnchor &&
+                   exprBeforeKeyword.GetDocumentEndOffset().Offset = tokenBeforeKeywordEnd then
+                   Some (indentAnchor.GetDocumentStartOffset().Offset)
+                else None
 
-                    | _ -> defaultTraverse expr }
-
-        let documentCoords = document.GetCoordsByOffset(lexer.TokenStart)
-        match SyntaxTraversal.Traverse(getPosFromCoords documentCoords, parseTree, visitor) with
+            match expr with
+            // if expr then ...
+            // elif ... then {caret}
+            | :? IElifExpr as elifExpr when (elifExpr.ThenExpr :? IFromErrorExpr) ->
```

Comment:
Should it limit to error expressions with zero length? Or do we expect it to work with erroneous expressions that could contain other ones?

What if we later decide to not convert empty FCS error expressions? I.e. the expression would be simply missing. Should we cover this case too?

Reply:
Initially, when I was rewriting the code, I kept the condition with ErrorExpr, and maybe some tests didn't pass without it. The code may have changed from then.

Now I tried to remove this restriction and the tests passed!  Even more than that, it fixes the old behavior for the case 
```fsharp
if true &&
        true then {caret}5
```
which resulted in
```fsharp
if true &&
        true then
          {caret}5
```

Added tests, thanks.